README.md
# krunker-logo-maker

## Site: https://hitthemoney.com/krunker-logo-maker


changelog.txt
1.4.2

Working On:
- Adding more settings
- Better text box
- More and better animations
- More export formats for animation
- Fixing Bugs

Update v1.4.2
- Bug Fixes
- Changed changelog CSS/Styles

Update v1.4.1
- Added more download options for the canvas method
- Bug Fixes

Update v1.4.0
- Added animations! (currently only WebM export is available)
- Fixed Bugs

Update v1.3.0
- Added new logo making method! (canvas)
- New method has hue setting
- Settings container can now be resized
- Added an option to remove the background animation

Update v1.2.4
- Added discord bot button

Update v1.2.3
- Fixed Bugs
- Added donate button

Update v1.2.2
- Fixed Bugs
- Gray text overlay is now a bit more smooth and less jagged

Update v1.2.1
- Added a text height offset setting

Update v1.2.0:
- Added a drop shadow text option
- Added the ability to download as SVG and JPEG (JPEG is not recommended)
- Added the ability to make "season 3" logos
- Redid the settings css

Update v1.1.1:
- Fixed png not updating after loading an icon

Update v1.1.0:
- Fixed hue bug
- Added stroke width (currently buggy in the preview)
- Added an icon image (replaces the skull on the logo with your own image)
- Added icon scale, icon positioning, and icon hue

Update v1.0.1:
- Bug fixes
- New UI (Thanks to Hoodgail!)

Update v1.0.0:
- Initial Release

js/svglogo.js
var usernamePathBG = d.getElementById("usernamePathBG"),
    usernamePath = d.getElementById("usernamePath"),
    input = d.getElementById("input"),
    pathBG = d.getElementById("pathBG"),
    card = d.getElementById("card"),
    textHolder = d.getElementById("textHolder"),
    slope = 0.15625,
    textPath = d.getElementById("textPath"),
    topWings = d.getElementById("topWings"),
    topWings2 = d.getElementById("topWings2"),
    iconInput = d.getElementById("iconInput"),
    krunkerWings = d.getElementById("krunkerWings"),
    krunkerWingsIcon = d.getElementById("krunkerWingsIcon");
window.hue = 0;
window.iconHue = 0;
window._3dVal = 50;
window.season = "2";
window.heightOffset = 0;

krunkerWings.href.animVal = wingsWithIcon
krunkerWings.href.baseVal = wingsWithIcon

updatePreview = (updateText) => {
    setTimeout(() => {
        usernamePath.textContent = input.value.toUpperCase();
        usernamePathBG.textContent = input.value.toUpperCase();
        if (input.value.split(" ").join("") == "") {
            usernamePath.textContent = "YOURNAME";
            usernamePathBG.textContent = "YOURNAME";
        }
        //document.getElementById("svg").getElementById("usernameBG").style.textShadow = "rgb(16, 16, 16) 0px " + (_3dVal / (_3dVal / 20)).toString() + "px" //document.getElementById("svg").getElementById("usernameBG").style.textShadow.replace(_3dVal.toString() + "px", (_3dVal / 2.5).toString() + "px");
        svgToImg(document.getElementById("svg"), input.value);
        //document.getElementById("svg").getElementById("usernameBG").style.textShadow = "rgb(16, 16, 16) 0px " + _3dVal.toString() + "px" //document.getElementById("svg").getElementById("usernameBG").style.textShadow.replace((Math.round((_3dVal / 2.5) * 10000) / 10000).toString() + "px", _3dVal.toString() + "px");
        if (updateText !== false) {
            let usernameBBox = usernamePath.getBBox();
            let width = usernameBBox.width
            let boxWidth = width + 5
            let topCurve = boxWidth / 2
            let startingPoint = 640 - topCurve
            if (width >= 1280) document.getElementById("svg").setAttribute("width", `${width + 25}`)
            else document.getElementById("svg").setAttribute("width", `1280`)
            if (boxWidth >= 1280) {
                let size = 0.85;
                textPath.setAttribute("d", `M 0 360 q ${width / 2} ${(width / 2) * slope} ${width} 0`)
                textHolder.setAttribute("transform", `translate(96 54) scale(${size})`)
                topWings.setAttribute("stroke-width", "20")
                d.getElementById("sizeSlider").value = size;
                d.getElementById("sizeInput").value = size;
            } else if (720 <= boxWidth && boxWidth <= 1280) {
                let size = 1;
                textHolder.setAttribute("transform", `translate(0 7) scale(${size})`)
                topWings.setAttribute("stroke-width", "20")
                d.getElementById("sizeSlider").value = size;
                d.getElementById("sizeInput").value = size;
            } else if (480 <= boxWidth && boxWidth <= 720) {
                let size = 1.15
                textHolder.setAttribute("transform", `translate(-96 -45) scale(${size})`)
                topWings.setAttribute("stroke-width", "20")
                d.getElementById("sizeSlider").value = size;
                d.getElementById("sizeInput").value = size;
            } else if (boxWidth <= 480) {
                let size = 1.3
                textHolder.setAttribute("transform", `translate(-192 -95) scale(${size})`)
                topWings.setAttribute("stroke-width", "30")
                d.getElementById("sizeSlider").value = size;
                d.getElementById("sizeInput").value = size;
            } else {
                textHolder.setAttribute("transform", "translate(0 7) scale(1)")
                topWings.setAttribute("stroke-width", "40")
                d.getElementById("sizeSlider").value = 1;
                d.getElementById("sizeInput").value = 1;
            }
            if (startingPoint >= 130) {
                let num = 0
                if (usernamePath.textContent.slice(0, 1).toLowerCase() == "j") num = 38
                boxWidth = width - 100 - num
                topCurve = boxWidth / 2
                startingPoint = 640 - topCurve
                pathBG.setAttribute("d",
                    `M ${startingPoint} ${245 + (topCurve * slope) / 2} q ${topCurve} -${topCurve * slope} ${1280 - startingPoint * 2} 0`
                )
                var height = (380 - (_3dVal / 5)) + window.heightOffset
                //console.log(height)
                textPath.setAttribute("d",
                    `M 0 ${height} q 640 -100 1280 0`
                )
                //`M 0 380 q 640 -100 1280 0`
                topWings.style.display = "none"
            } else {
                topWings.style.display = ""
                topWings.setAttribute("d",
                    `M 130 350 q 445 -69.53125 1022 0`
                )
                boxWidth = width - 100
                topCurve = boxWidth / 2
                startingPoint = 640 - topCurve
                pathBG.setAttribute("d",
                    `M ${startingPoint} ${225 + (topCurve * slope) / 2} q ${topCurve} -${topCurve * slope} ${1280 - startingPoint * 2} 0`
                )
                var height = (360 - (_3dVal / 5)) + window.heightOffset
                textPath.setAttribute("d",
                    `M 0 ${height} q 640 -100 1280 0`
                )
            }
            boxWidth = width + 5
            if (boxWidth >= 1280) {
                topWings2.style.display = ""
                textPath.setAttribute("d",
                    `M -${(width - 1280) / 2} 360 q ${width / 2} -${(width / 2) * slope} ${width} 0`
                )
            } else {
                topWings2.style.display = "none"
            }
        }
    }, 1);

}

changeHue = (hue = d.getElementById("hueInput").value, updateInputs = false) => {
    //setTimeout(() => {
    if (updateInputs) {
        d.getElementById("hueInput").value = hue;
        d.getElementById("hueSlider").value = hue;
    }
    window.hue = hue;
    d.getElementById("svg").style =
        `transform: scale(0.5) translate(0%, -50%); filter: hue-rotate(${hue}deg)`
    krunkerWingsIcon.style.filter =
        `hue-rotate(${window.iconHue - window.hue}deg)`
    svgToImg(d.getElementById("svg"), input.value);
    //}, 1);
}

changeStroke = () => {
    //setTimeout(() => {
    window.strokeW = d.getElementById("strokeInput").value
    d.getElementById("usernameBG").setAttribute("stroke-width", `${strokeW * 2}px`)
    svgToImg(d.getElementById("svg"), input.value);
    //}, 1);
}

changeSize = () => {
    setTimeout(() => {
        window.size = d.getElementById("sizeInput").value
        var x,
            y;
        if (size > 1) {
            x = (-(((size - 1) * 1280) / 2))
            y = (-(((size - 1) * 720) / 2)) + 13
        } else if (size < 1) {
            x = (((1 - size) * 1280) / 2)
            y = (((1 - size) * 720) / 2) + 7
        } else if (size == 1) {
            x = 0
            y = 7
        }
        textHolder.setAttribute("transform", `translate(${x} ${y}) scale(${size})`)
        svgToImg(d.getElementById("svg"), input.value);
        let usernameBBox = usernamePath.getBBox();
        let width = usernameBBox.width * size
        if (width >= 1280) document.getElementById("svg").setAttribute("width", `${width + 50}`)
        else document.getElementById("svg").setAttribute("width", `1280`)
    }, 1);
}

function svgToImg(svg, name) {
    svg.style = `transform: scale(1); filter: hue-rotate(${hue}deg)`;
    document.getElementById("svg").getElementById("usernameBG").style.textShadow = "rgb(16, 16, 16) 0px " + (_3dVal / 2).toString() + "px"
    var svgData = new XMLSerializer().serializeToString(svg);
    document.getElementById("svg").getElementById("usernameBG").style.textShadow = "rgb(16, 16, 16) 0px " + (_3dVal).toString() + "px"
    //document.getElementById("svg").getElementById("usernameBG").style.textShadow = "rgb(16, 16, 16) 0px " + _3dVal.toString() + "px" //document.getElementById("svg").getElementById("usernameBG").style.textShadow.replace((Math.round((_3dVal / 2.5) * 10000) / 10000).toString() + "px", _3dVal.toString() + "px");
    var svgData2 = new XMLSerializer().serializeToString(svg);
    svg.style = `transform: scale(0.5) translate(0%, -50%); filter: hue-rotate(${hue}deg`;

    window.canvas = d.getElementById("canvas") //createElement("canvas");
    canvas.width = svg.getAttribute('width');
    canvas.height = svg.getAttribute('height');
    var ctx = canvas.getContext("2d");

    window.img = d.createElement("img");
    let svgDataUrl = "data:image/svg+xml;base64," + btoa(unescape(encodeURIComponent(svgData)));
    img.setAttribute("src", svgDataUrl);
    img.id = "previewImg"

    img.onload = function () {
        let hue2 = hue
        if (hue2 == 360) hue2 = 720
        //ctx.filter = 'hue-rotate(' + hue2 / 2 + 'deg)';
        ctx.drawImage(img, 0, 0);
        pngLink = canvas.toDataURL("image/png");
        let imgSrcElem = d.getElementsByClassName("imgSrc")
        for (num = 0; num < imgSrcElem.length; num++) imgSrcElem[num]["content"] = pngLink
        d.getElementById("downloadPng").download = `${name}'s_Krunker_Logo.PNG`;
        d.getElementById("downloadPng").href = pngLink;

        d.getElementById("downloadJpg").download = `${name}'s_Krunker_Logo.JPG`;
        d.getElementById("downloadJpg").href = canvas.toDataURL("image/jpeg");

        d.getElementById("downloadSvg").download = `${name}'s_Krunker_Logo.SVG`;
        d.getElementById("downloadSvg").href = "data:image/svg+xml;base64," + btoa(unescape(encodeURIComponent(svgData2)));
    };
};

function textBox(checked) {
    if (checked == true) {
        pathBG.style.display = "";
    } else if (checked == false) {
        pathBG.style.display = "none";
    };
}

iconInput.addEventListener('change', function () {
    changeIcon(this.files);
}, false);

changeIcon = (files) => {
    let icon = files[0]
    //let iconUrl = URL.createObjectURL(icon)
    blobToDataURL(icon, function (dataUrl) {
        //console.log(dataUrl);
        d.getElementById("iconSInput").style.cursor = "initial";
        d.getElementById('iconSSlider').style.cursor = "initial";
        d.getElementById("iconHueInput").style.cursor = "initial";
        d.getElementById('iconHueSlider').style.cursor = "initial";
        d.getElementById("iconCInputX").style.cursor = "initial";
        d.getElementById('iconCInputY').style.cursor = "initial";
        window.iconChanged = true
        if (season == "2") {
            krunkerWings.href.baseVal = wingsNoIcon
            krunkerWings.href.animVal = wingsNoIcon
        } else if (season == "3") {
            krunkerWings.href.animVal = wingsS3NoIcon
            krunkerWings.href.baseVal = wingsS3NoIcon
        }
        krunkerWingsIcon.style.display = ""
        krunkerWingsIcon.src = dataUrl
        svgToImg(document.getElementById("svg"), input.value);
        //krunkerWingsIcon.href.animVal = dataUrl
        //krunkerWingsIcon.href.baseVal = dataUrl
    });
}

changeIconScale = () => {
    let value = d.getElementById("iconSInput").value
    if (window.iconChanged == true) {
        /*let x = 541.5 * (1 - (value - 1)) //541.5 - (value * 100 * (1 + ((value - 1)/2)));
        console.log(x)*/
        let y = 240 + (100 * (1 + (1 - value))); //240 + (100 * ((1 + (1 - value)) * value)); //240 + (100 * (1 + (1 - value))); //340
        if (value > 1) {
            y = 240 + (100 * ((1 + (1 - value)) * value));
        }
        krunkerWingsIcon.style.transform = `scale(${value})`
        krunkerWingsIcon.style.transformOrigin = "center"
        //krunkerWingsIcon.setAttribute("x", x)
        d.getElementById('iconCInputY').value = y;
        krunkerWingsIcon.setAttribute("y", y)
        svgToImg(document.getElementById("svg"), input.value);
    } else {
        d.getElementById("iconSInput").value = 1;
        d.getElementById('iconSSlider').value = 1;
    }
}

changeIconCords = (xy) => {
    if (window.iconChanged == true) {
        if (xy == "x") {
            //krunkerWingsIcon.setAttribute("x", d.getElementById('iconCInputX').value)
            krunkerWingsIcon.style.left = d.getElementById('iconCInputX').value + "px"
        } else if (xy == "y") {
            //krunkerWingsIcon.setAttribute("y", d.getElementById('iconCInputY').value)
            krunkerWingsIcon.style.top = d.getElementById('iconCInputY').value + "px"
        }
        svgToImg(document.getElementById("svg"), input.value);
    } else {
        d.getElementById('iconCInputX').value = 541.5
        d.getElementById('iconCInputY').value = 340
    }
}

changeIconHue = () => {
    if (window.iconChanged == true) {
        window.iconHue = d.getElementById("iconHueInput").value
        krunkerWingsIcon.style.filter = `hue-rotate(${iconHue - window.hue}deg)`
        svgToImg(document.getElementById("svg"), input.value);
    } else {
        d.getElementById("iconHueInput").value = 0;
        d.getElementById("iconHueSlider").value = 0;
    }
}

update3dWidth = (id) => {
    setTimeout(() => {
        var elem = document.getElementById(id)
        document.getElementById("svg").getElementById("usernameBG").style.textShadow = "rgb(16, 16, 16) 0px " + (elem.value).toString() + "px" //document.getElementById("svg").getElementById("usernameBG").style.textShadow.replace(_3dVal.toString() + "px", (elem.value).toString()  + "px");
        window._3dVal = elem.value;
    }, 1);
}

downloadImage = (format) => {
    document.getElementById("download" + format).click();
}

function changeSeason(val) {

    //krunkerWings.href.baseVal = wingsNoIcon
    //krunkerWings.href.animVal = wingsNoIcon

    window.season = val;
    if (val == "2") {
        if (krunkerWingsIcon.src == document.URL) {
            krunkerWings.href.animVal = wingsWithIcon
            krunkerWings.href.baseVal = wingsWithIcon
        } else {
            krunkerWings.href.animVal = wingsNoIcon
            krunkerWings.href.baseVal = wingsNoIcon
        }
    } else if (val == "3") {
        if (krunkerWingsIcon.src == document.URL) {
            krunkerWings.href.animVal = wingsS3WithIcon
            krunkerWings.href.baseVal = wingsS3WithIcon
        } else {
            krunkerWings.href.animVal = wingsS3NoIcon
            krunkerWings.href.baseVal = wingsS3NoIcon
        }
    }
}

updateHeightOffset = (id) => {
    setTimeout(() => {
        var elem = document.getElementById(id);
        window.heightOffset = (elem.value) * -1;
        document.getElementById("svg").getElementById("textOffsetHolder").style.transform = `translateY(${heightOffset}px)` //bottom = `${elem.value}px`
    }, 1);
}

js/animation.js
let isPowerOf2 = (value) => {
    return (value & (value - 1)) == 0;
}

let animCanvas = document.getElementById("animationCanvas"),
    animVideo = document.getElementById("animationResult");

class CanvasRecorder {
    /**
     * Canvas Renderer/Recorder
     * @param {HTMLCanvasElement} canvas 
     * @param {HTMLVideoElement} video
     * Some of this is from https://github.com/webrtc/samples/blob/gh-pages/src/content/capture/canvas-record/js/main.js
     */
    constructor(canvas, video) {
        this.canvas = canvas;
        this.video = video;

        this.canvas.style.display = "";
        this.video.style.display = "none";

        this.downloadName = "Krunker_Logo_Animation";
        this.mediaRecorder;
        this.recordedBlobs;
        this.sourceBuffer;
        this.mediaSource = new MediaSource();
        this.mediaSource.addEventListener("sourceopen", () => {
            this.sourceBuffer = this.mediaSource.addSourceBuffer("video/webm; codecs=\"vp8\"");
        }, false);
        this.stream = this.canvas.captureStream();
    }

    startRecording() {
        this.canvas.style.display = "inline";
        this.video.style.display = "none";

        this.video.controls = false;
        let options = {
            mimeType: "video/webm"
        };
        this.recordedBlobs = [];
        this.mediaRecorder = new MediaRecorder(this.stream, options);
        try {
            this.mediaRecorder = new MediaRecorder(this.stream, options);
        } catch (err0) {
            console.log("Unable to create MediaRecorder with options Object: ", err0);
            try {
                options = {
                    mimeType: "video/webm,codecs=vp9"
                };
                this.mediaRecorder = new MediaRecorder(this.stream, options);
            } catch (err1) {
                console.log("Unable to create MediaRecorder with options Object: ", err1);
                try {
                    options = "video/vp8"; // Chrome 47
                    this.mediaRecorder = new MediaRecorder(this.stream, options);
                } catch (err2) {
                    alert(`\
MediaRecorder is not supported by this browser.

Try Firefox 29 or later, or Chrome 47 or later, with Enable experimental Web Platform features enabled from chrome://flags.\
`);
                    console.error("Exception while creating MediaRecorder:", err2);
                    return;
                }
            }
        }

        this.mediaRecorder.onstop = (event) => {
            const superBuffer = new Blob(this.recordedBlobs, {
                type: "video/webm"
            });
            this.video.src = window.URL.createObjectURL(superBuffer);
        };

        this.mediaRecorder.ondataavailable = (event) => {
            if (event.data && event.data.size > 0) {
                this.recordedBlobs.push(event.data);
            }
        };

        this.mediaRecorder.start(100); // collect 100ms of data

        console.log("Started MediaRecorder")
    }

    stopRecording() {
        this.mediaRecorder?.stop();

        this.canvas.style.display = "none";
        this.video.style.display = "";
        this.video.controls = true;

        console.log("Stopped MediaRecorder")
    }

    download() {
        const blob = new Blob(this.recordedBlobs, {
            type: "video/webm"
        });
        const url = window.URL.createObjectURL(blob);
        const a = document.createElement("a");
        a.style.display = "none";
        a.href = url;
        a.download = this.downloadName + ".webm";
        document.body.appendChild(a);
        a.click();
        setTimeout(() => {
            document.body.removeChild(a);
            window.URL.revokeObjectURL(url);
        }, 50);
    }
}

class LogoAnimation {
    /**
     * Krunker Logo Animation
     * @param {HTMLCanvasElement} canvas 
     * @param {HTMLImageElement} image 
     */
    constructor(canvas, image) {
        this.canvas = canvas;
        this.image = image;
        this.gl = canvas.getContext("webgl");
        this.frame = 0;
        this.isAnimating = false;
        this.onAnimationFinished = () => {};
    }

    init() {
        const {
            gl,
            image,
            canvas
        } = this;

        // this.updateTexture();

        let vertices = [
            -1, 1, 0.0,
            -1, -1, 0.0,
            1, -1, 0.0,
            1, 1, 0.0
        ];

        let indices = [3, 2, 1, 3, 1, 0];

        let vertexBuffer = gl.createBuffer()
        gl.bindBuffer(gl.ARRAY_BUFFER, vertexBuffer);
        gl.bufferData(gl.ARRAY_BUFFER, new Float32Array(vertices), gl.STATIC_DRAW);
        gl.bindBuffer(gl.ARRAY_BUFFER, null);

        let indexBuffer = gl.createBuffer();
        gl.bindBuffer(gl.ELEMENT_ARRAY_BUFFER, indexBuffer);
        gl.bufferData(gl.ELEMENT_ARRAY_BUFFER, new Uint16Array(indices), gl.STATIC_DRAW);
        gl.bindBuffer(gl.ELEMENT_ARRAY_BUFFER, null);

        const vertexShaderSource = `
        attribute vec3 coordinates;
        varying vec2 _coordinates; 
        void main() {
            _coordinates = vec2(coordinates.x, coordinates.y);
            gl_Position = vec4(coordinates, 1.0);
        }
        `;

        const fragShaderSource = `
        precision highp float;

        // Defines
        #define PI ${Math.PI}

        // Globals
        varying vec2 _coordinates; 

        // Uniforms
        uniform float uFrame; // frame count
        uniform vec2 uRes; // canvas resolution
        uniform sampler2D uSampler;
        
        mat2 rotate2d(float angle) {
            return mat2(cos(angle), -sin(angle), sin(angle), cos(angle));
        }

        void main() {
            float movWidth = 1.25;
            vec2 uv = gl_FragCoord.xy / uRes.xy; // _coordinates;
            vec2 texUv = uv;
            texUv.y = 1.0 - uv.y;
            vec4 tex = texture2D(uSampler, texUv);
            float time = sin((uFrame / 60.) - movWidth) / (2. / movWidth);
            uv -= 0.5;
            uv = rotate2d((-10. / 360.) * PI) * uv; 
            
            vec3 col = tex.rgb;
            vec3 shine = clamp(vec3( (1.0) - sqrt(12.5 * (time == uv.x ? 0.0 : time > uv.x ? time - uv.x : uv.x - time) + 0.10) ), 0.0, 1.0);
            col = mix(col, vec3(1.0), shine);

            // Output to screen
            gl_FragColor = vec4(col, tex.a);
        }
        `

        /**
         * For creating WebGL Shaders
         * @param {string} sourceCode GLSL Shader Source
         * @param {35632|35633} type WebGL Shader Type
         * @tutorial https://developer.mozilla.org/en-US/docs/Web/API/WebGLShader
         * @returns {WebGLShader}
         */
        function createShader(sourceCode, type) {
            // Compiles either a shader of type gl.VERTEX_SHADER or gl.FRAGMENT_SHADER
            var shader = gl.createShader(type);
            gl.shaderSource(shader, sourceCode);
            gl.compileShader(shader);

            if (!gl.getShaderParameter(shader, gl.COMPILE_STATUS)) {
                var info = gl.getShaderInfoLog(shader);
                throw "Could not compile WebGL program. \n\n" + info;
            }
            return shader;
        }

        const shaderProgram = gl.createProgram();

        const fragmentShader = createShader(fragShaderSource, gl.FRAGMENT_SHADER);
        const vertexShader = createShader(vertexShaderSource, gl.VERTEX_SHADER);

        gl.attachShader(shaderProgram, vertexShader);
        gl.attachShader(shaderProgram, fragmentShader);

        gl.linkProgram(shaderProgram);

        let progLog = gl.getProgramInfoLog(shaderProgram);
        if (!!progLog) console.log(progLog)

        gl.bindBuffer(gl.ARRAY_BUFFER, vertexBuffer);
        gl.bindBuffer(gl.ELEMENT_ARRAY_BUFFER, indexBuffer);
        let coord = gl.getAttribLocation(shaderProgram, "coordinates");

        this.animate = () => {
            let frameLocation = gl.getUniformLocation(shaderProgram, "uFrame"),
                samplerLocation = gl.getUniformLocation(shaderProgram, "uSampler"),
                resLocation = gl.getUniformLocation(shaderProgram, "uRes");

            gl.activeTexture(gl.TEXTURE0);
            // Bind the texture to texture unit 0
            gl.bindTexture(gl.TEXTURE_2D, this.texture);

            gl.vertexAttribPointer(coord, 3, gl.FLOAT, false, 0, 0);
            gl.enableVertexAttribArray(coord);

            gl.useProgram(shaderProgram);

            gl.uniform1f(frameLocation, this.frame);
            gl.uniform2fv(resLocation, [canvas.width, canvas.height]);
            // Tell the shader we bound the texture to texture unit 0
            gl.uniform1i(samplerLocation, 0);

            let height = canvas.height / (canvas.width > canvas.height ? (canvas.height / canvas.width) : 1);
            let width = canvas.width * (canvas.width < canvas.height ? (canvas.height / canvas.width) : 1);

            gl.viewport(0, 0, width, height) //, gl.drawingBufferWidth, gl.drawingBufferHeight);
            gl.clearColor(1.0, 1.0, 1.0, 0.0)
            gl.clear(gl.COLOR_BUFFER_BIT);

            gl.drawElements(gl.TRIANGLES, indices.length, gl.UNSIGNED_SHORT, 0);

            if (this.frame > 60 * 6) {
                this.onAnimationFinished();
            } 

            this.frame++;
        }
    }

    updateTexture() {
        const {
            gl,
            image
        } = this;

        // https://developer.mozilla.org/en-US/docs/Web/API/WebGL_API/Tutorial/Using_textures_in_WebGL

        const imgLevel = 0;
        const internalFormat = gl.RGBA;
        const srcFormat = gl.RGBA;
        const srcType = gl.UNSIGNED_BYTE;

        this.texture = gl.createTexture();
        // gl.bindTexture(gl.TEXTURE_2D, this.texture);

        gl.bindTexture(gl.TEXTURE_2D, this.texture);
        gl.texImage2D(gl.TEXTURE_2D, imgLevel, internalFormat, srcFormat, srcType, image);

        if (isPowerOf2(image.width) && isPowerOf2(image.height)) {
            // Yes, it's a power of 2. Generate mips.
            gl.generateMipmap(gl.TEXTURE_2D);
        } else {
            // No, it's not a power of 2. Turn off mips and set
            // wrapping to clamp to edge
            gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_S, gl.CLAMP_TO_EDGE);
            gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_T, gl.CLAMP_TO_EDGE);
            gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.LINEAR);
        }
    }

    startAnimation() {
        this.isAnimating = true;
        // let animationTimeout;

        let animationLoop = () => {
            if (this.isAnimating) {
                this.animate();
                // animationTimeout = setTimeout(animationLoop, 1000 / 60);
                requestAnimationFrame(animationLoop);
            } else {
                // clearTimeout(animationTimeout);
                cancelAnimationFrame(animationLoop)
            }
        }

        animationLoop();
    }

    stopAnimation() {
        this.isAnimating = false;
    }
}

const recorder = new CanvasRecorder(animCanvas, animVideo);
const animator = new LogoAnimation(animCanvas, new Image(window.smallImgB64));
animator.init();

let downloadAnimation = () => {
    recorder.download();
}

let updateAnimationScale = (width, height) => {
    let fn = (elem) => {
        elem.style.height = height / 2 + "px";
        elem.style.width = width / 2 + "px";
        elem.setAttribute("height", height);
        elem.setAttribute("width", width);
    }

    [animVideo, animCanvas].forEach(fn);
}

let stopRenderAnimation = () => {
    recorder.stopRecording();
    animator.stopAnimation();
}

window.stopRenderAnimation = stopRenderAnimation;
animator.onAnimationFinished = stopRenderAnimation;

let renderAnimation = () => {
    let pngDownload = document.getElementById("downloadPng");
    recorder.downloadName = pngDownload.download.replace(/(\.\w+)$/, "");
    let imgUrl = pngDownload.href;
    let img = new Image();
    img.src = imgUrl;
    img.onload = (event) => {
        updateAnimationScale(img.width, img.height);
        animator.image = img;
        animator.updateTexture();
        animator.frame = 0;
        animator.startAnimation();
        recorder.startRecording();
    }
}

let showAnimationHolder = () => {
    let animationHolder = document.getElementById("animationHolder");
    animationHolder.style.display = "block";
    document.getElementById("downloadHolder").style.display = "none";
    document.getElementById("popupHolder").style.display = "block";
}

js/index.js
var input = d.getElementById("input"),
    settingsH = d.getElementById("settingsHeader"),
    settings = d.getElementById("settings"),
    bgAnimCheck = d.getElementById("bgAnimCheck"),
    svgSettings = d.getElementById("svgSettings"),
    canvasSettings = d.getElementById("canvasSettings"),
    svgLogo = d.getElementById("svg"),
    canvasLogo = d.getElementById("canvasLogo"),
    downloadBtn = d.getElementById("downloadBtn"),
    downloadBtnOnclick = `updatePreview(false); showDownloadPopup();`,
    bottomResizer = d.getElementById("bottomSettResizer"),
    settResizers = d.querySelector("#settResizers"),
    settContainer = d.getElementById("settContainer");

window.settingsClosed = false;
window.popups = ["changelog", "download", "animation"];
window.bgAnim = (localStorage.getItem("bgAnim") === "false") ? false : true;
window.logoType = "svg";

;

(async function () {
    var version = document.getElementById("version"),
        changelog = await fetch("./changelog.txt"),
        changelogText = await changelog.text(),
        changelog2 = document.getElementById("changelog")
    version.innerHTML = changelogText.slice(0, 6)
    changelog2.innerHTML = changelogText
})();

svgToImg(document.getElementById("svg"), "YOURNAME");
window.canvasLm = new LogoMaker(canvasLogo);
canvasLogo.style.transform = "scale(0.5) translate(0%, -50%)";
canvasLm.updateLogo();

function checkSettings(event) {
    switch (event) {
        case "out":
            if (settingsClosed == false) {
                settingsH.innerHTML = "Settings";
            }
            break;
        case "hover":
            if (settingsClosed == false) {
                settingsH.innerHTML = "Close Settings";
            }
            break;
        case "click":
            if (settingsClosed == false) {
                settingsClosed = true;
                settings.style = "height: 78px; overflow: hidden;";
                settingsH.innerHTML = "Show Settings";
            } else if (settingsClosed == true) {
                settingsClosed = false;
                settings.style = (logoType === "svg") ? "" : `height: ${settContainer.getBoundingClientRect().height}px`;
                settingsH.innerHTML = "Settings";
            }
            break;
    }
}

function showChangelog() {
    let changelogHolder = document.getElementById("changelogHolder");
    changelogHolder.style.display = "block";
    document.getElementById("popupHolder").style.display = "block";
}

function hidePopup() {
    clearInterval(window.interval);
    for (let num = 0; num < popups.length; num++) document.getElementById(popups[num] + "Holder").style.display = "none"
    document.getElementById("popupHolder").style.display = "none"
    if (window.stopRenderAnimation) window.stopRenderAnimation();
    document.getElementById("bottom").style.zIndex = "";
}

function showDownloadPopup() {
    let downloadHolder = document.getElementById("downloadHolder");
    downloadHolder.style.display = "block";
    document.getElementById("popupHolder").style.display = "block";
    document.getElementById("bottom").style.zIndex = -9999;
}

function blobToDataURL(blob, callback) {
    var fr = new FileReader();
    fr.onload = function (e) {
        callback(e.target.result);
    }
    fr.readAsDataURL(blob);
}

toggleBgAnim = (check) => {
    window.bgAnim = check;
    localStorage.setItem("bgAnim", check);
    if (check) {
        document.body.style = "animation-play-state: running;"
    } else {
        document.body.style = "animation-play-state: paused;"
    }
}

/*changeSlope = () => {
    slope = d.getElementById('slopeInput').value;
    updatePreview();
}*/

if (!window.bgAnim) {
    bgAnimCheck.style = "transition: none;";
    bgAnimCheck.click();
    bgAnimCheck.style = "";
}

toggleBgAnim(window.bgAnim);

changeLogoType = (val) => {
    window.logoType = val;
    if (val === "svg") {
        svgSettings.style.display = "block";
        canvasSettings.style.display = "none";
        settings.style.height = "";
        downloadBtn.innerHTML = "Download Logo";
        changeHue(canvasLm.hue, true);
        updatePreview();
        canvasLogo.style.opacity = "0";
        svgLogo.style.opacity = "1";
        setTimeout(() => {
            canvasLogo.style.display = "none";
            svgLogo.style.display = "";
        }, 500)
        downloadBtn.onclick = () => {
            eval(downloadBtnOnclick);
        };
        d.getElementById("downloadSvgBtn").style.display = ""
        input.onkeydown = updatePreview;
    } else if (val === "canvas") {
        canvasSettings.style.display = "block";
        svgSettings.style.display = "none";
        settings.style.height = settContainer.getBoundingClientRect().height + "px";
        downloadBtn.innerHTML = "Download Logo";
        d.getElementById("downloadSvgBtn").style.display = "none"
        downloadBtn.onclick = () => {
            // window.canvasLm.download();
            window.showDownloadPopup();
            canvasLm.updateDownloads();
        }
        input.onkeydown = () => {};

        let a = b => d.getElementById(b);

        a("canvasHueInput").value = window.hue;
        a("canvasHueSlider").value = window.hue;
        canvasLm.changeValue(input.value);
        canvasLm.changeHue(window.hue);
        canvasLm.updateLogo();
        canvasLogo.style.opacity = "1";
        svgLogo.style.opacity = "0";
        setTimeout(() => {
            canvasLogo.style.display = "";
            svgLogo.style.display = "none";
        }, 500);
    }
}

makeSettingsResizable = () => {
    bottomResizer.addEventListener("mousedown", function (e) {
        e.preventDefault();
        window.addEventListener("mousemove", resize);
        window.addEventListener("mouseup", stopResize);
    })

    function resize(e) {
        settResizers.style.pointerEvents = "all";
        document.body.style.cursor = "ns-resize";
        settings.style.transition = "none";
        settings.style.height = `${e.pageY}px`;
    }

    function stopResize() {
        settResizers.style.pointerEvents = "";
        document.body.style.cursor = "";
        settings.style.transition = "";
        window.removeEventListener("mousemove", resize)
    }
}

makeSettingsResizable(".resizable")